core/src/main/scala/org/apache/spark/sql/execution/CompactExecRowToMutableRow.scala

```diff
@@ -221,6 +223,6 @@ trait CompactExecRowToMutableRow extends ResultWasNull {
       }
       i = i + 1
     }
-    mutableRow
+    unsafeproj.apply(mutableRow)
```

Comment:
This is unnecessary for the call from CachedBatchCreator. I will suggest this one always returning SpecificMutableRow while the child classes apply the unsafe projection if required (i.e. only CompactExecRowIteratorOnRS will while CachedBatchCreator will not).

Ideally this whole class can be changed to directly create UnsafeRow directly (preferably with CodeGeneration) instead of SpecificMutableRow, but since the main caller is RowFormatScanRDD its not that important (esp for TPCH) while it will probably not make a lot of difference for CachedBatchCreator once the change as I suggested is done.